sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -86,6 +86,7 @@ abstract class Optimizer(catalogManager: CatalogManager)
         PushDownPredicates,
         PushDownLeftSemiAntiJoin,
         PushLeftSemiLeftAntiThroughJoin,
+        EliminateSorts,
```

Comment:
this means we depend on the rule order, which violates the design goal of catalyst.

Can we update the `EliminateSorts` to allow `LocalLimit` between Join and Sort?

Reply:
Eliminate the sorts between `Join` and `Sort` may affect the results, for example:
https://github.com/apache/spark/blob/e9cc1024df4d587a0f456842d495db91984ed9db/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala#L366-L376
```
== FAIL: Plans do not match ===
 'GlobalLimit 10                                'GlobalLimit 10
 +- 'LocalLimit 10                              +- 'LocalLimit 10
    +- 'Join LeftOuter                             +- 'Join LeftOuter
       :- LocalLimit 10                               :- LocalLimit 10
!      :  +- Project [a#0, b#0]                       :  +- Sort [a#0 ASC NULLS FIRST, b#0 DESC NULLS LAST], true
!      :     +- LocalRelation [a#0, b#0, c#0]         :     +- Project [a#0, b#0]
!      +- Project [d#0]                               :        +- LocalRelation [a#0, b#0, c#0]
!         +- LocalRelation [d#0]                      +- Project [d#0]
!                                                        +- LocalRelation [d#0]
```

Reply:
eliminating sort under join will definitely change result, even before this PR, the point is, do we need to guarantee the ordering according to the query semantic.

Reply:
OK. Updated the code.